File: ai-service/app/providers/llm.py
import json
import os

from dotenv import load_dotenv
from openai import OpenAI

load_dotenv()

_client = OpenAI(api_key=os.environ["OPEN_AI_API_KEY"])
_MODEL = os.getenv("RESUME_PARSER_MODEL","gpt-4o-mini")

def complete_resume_json(*,system_prompt:str,resume_text:str)-> dict:
    """
    Ask the model for a JSON object that matches ParsedResumeData.
    response_format json_object stops markdown fences around the JSON.
    """
    completion = _client.chat.completions.create(
        model=_MODEL,
        response_format={"type":"json_object"},
        temperature=0,
        messages=[
            {"role":"system",
            "content":system_prompt
             },
            {
                "role":"user",
                "content": f"Resume text: \n\n{resume_text}",
            },
        ]
    )

    content = completion.choices[0].message.content
    if not content:
        raise ValueError("LLm returned empty resume JSON")

    data = json.loads(content)
    if not isinstance(data,dict):
        raise ValueError("LLM JSON must be an object")

    return data

File: backend/src/ai/providers/llm.ts
import { AppError } from "../../errors/app-error";
import { env } from "../../config/env";
import { openai } from "./openai";

export async function completeJson(params:{
    systemPromt:string,
    userPrompt:string,
    model:string
}):Promise<Record<string,unknown>>{
    const completion = await openai.chat.completions.create({
        model:params.model ?? env.RESUME_PARSER_MODEL,
        response_format:{type:"json_object"},
        temperature:0,
        messages:[
            {role:"system",content:params.systemPromt},
            {role:"user",content:params.userPrompt}
        ],
    })

    const content = completion.choices[0]?.message.content

    if(!content){
        throw new AppError("LLM returned empty JSON",502)
    }

    let parsed:unknown
    try {
        parsed = JSON.parse(content)
    } catch (error) {
        throw new AppError("LLM returned invalid JSON",502)
    }
    if(!parsed || typeof parsed !== "object" || Array.isArray(parsed)){
        throw new AppError("LLM JSON must be an object",502)
    }
    return parsed as Record<string,unknown>
}